main.go

```diff
@@ -440,15 +456,15 @@ func main() {
 
 		ins := signalhttp.NewHandlerInstrumenter(reg, []string{"group", "handler"})
 
-		r.Group(func(r chi.Router) {
-			tenantIDs := map[string]string{}
-			authorizers := map[string]rbac.Authorizer{}
-			var rateLimits []ratelimit.Config
-			// registrationRetryCount used by authenticator providers to count
-			// registration failures per tenant.
-			registerTenantsFailingMetric := authentication.RegisterTenantsFailingMetric(reg)
-			pm := authentication.NewProviderManager(logger, registerTenantsFailingMetric)
+		tenantIDs := map[string]string{}
+		authorizers := map[string]rbac.Authorizer{}
+		var rateLimits []ratelimit.Config
+		// registrationRetryCount used by authenticator providers to count
+		// registration failures per tenant.
+		registerTenantsFailingMetric := authentication.RegisterTenantsFailingMetric(reg)
+		pm := authentication.NewProviderManager(logger, registerTenantsFailingMetric)
 
+		r.Group(func(r chi.Router) {
```

Comment:
> Looks like this could be moved even further down around line 515, since the tenant ID population logic etc. is not related to the router.

@matej-g `main()` is 900 lines long!  I am nervous about refactoring it as part of this PR.  You missed the `r` at line 505.  It wasn't obvious to me how the looping through `tenant` and pushing the `r.Group()` relate.  I think the refactor should be done -- other people will also struggle with this code -- but perhaps not as part of my first PR.

Reply:
You're right, this requires more thorough refactoring and that's beyond this PR, let's leave it aside.